packages/flutter_driver/lib/src/driver/refresh_rate_summarizer.dart

```diff
@@ -0,0 +1,138 @@
+  final List<TimelineEvent> vsyncEvents;
+
+  /// Computes a [RefreshRateSummary] over the period of the timeline.
+  RefreshRateSummary computeRefreshRateSummary() {
+    return RefreshRateSummary._(refreshRates: _computeRefreshRates());
+  }
+
+  List<double> _computeRefreshRates() {
+    final List<double> result = <double>[];
+    for (int i = 0; i < vsyncEvents.length; i++) {
+      final TimelineEvent event = vsyncEvents[i];
+      if (event.phase != 'B') {
+        continue;
+      }
+      assert(event.name == kUIThreadVsyncProcessEvent);
+      assert(event.arguments != null);
+      final Map<String, dynamic> arguments = event.arguments!;
+      const double nanosecondsPerSecond = 1e+9;
+      final int startTimeInNanoseconds = arguments['StartTime'] as int;
+      final int targetTimeInNanoseconds= arguments['TargetTime'] as int;
+      final int frameDurationInNanoseconds = targetTimeInNanoseconds - startTimeInNanoseconds;
+      final double refreshRate = nanosecondsPerSecond /
+          frameDurationInNanoseconds;
+      result.add(refreshRate);
+    }
+    return result;
+  }
+}
+
+/// A description of the summary generated by [RefreshRateSummarizer].
+class RefreshRateSummary {
+  RefreshRateSummary._({required List<double> refreshRates}) {
+    _numberOfTotalFrames = refreshRates.length;
+    for (final double refreshRate in refreshRates) {
+      if ((refreshRate - 30).abs() < _kErrorMargin) {
+        _numberOf30HzFrames++;
+        continue;
+      }
+      if ((refreshRate - 60).abs() < _kErrorMargin) {
+        _numberOf60HzFrames++;
+        continue;
+      }
+      if ((refreshRate - 90).abs() < _kErrorMargin) {
+        _numberOf90HzFrames++;
+        continue;
+      }
+      if ((refreshRate - 120).abs() < _kErrorMargin) {
+        _numberOf120HzFrames++;
+        continue;
+      }
+      _framesWithIllegalRefreshRate.add(refreshRate);
+    }
+    assert(_numberOfTotalFrames ==
+        _numberOf30HzFrames +
+            _numberOf60HzFrames +
+            _numberOf90HzFrames +
+            _numberOf120HzFrames +
+            _framesWithIllegalRefreshRate.length);
+  }
+
+  static const double _kErrorMargin = 6.0;
+
+  /// Number of frames with 30hz refresh rate
+  int get numberOf30HzFrames => _numberOf30HzFrames;
+
+  /// Number of frames with 60hz refresh rate
+  int get numberOf60HzFrames => _numberOf60HzFrames;
+
+  /// Number of frames with 90hz refresh rate
+  int get numberOf90HzFrames => _numberOf90HzFrames;
+
+  /// Number of frames with 120hz refresh rate
+  int get numberOf120HzFrames => _numberOf120HzFrames;
+
+  /// The percentage of 30hz frames.
+  ///
+  /// For example, if this value is 20, it means there are 20 percent of total
+  /// frames are 30hz. 0 means no frames are 30hz, 100 means all frames are 30hz.
+  double get percentageOf30HzFrames => _numberOfTotalFrames > 0
+      ? _numberOf30HzFrames.toDouble() / _numberOfTotalFrames * 100
```

Comment:
Are these toDouble() calls needed? I thought that "/" returned a double by default.

Reply:
No, that would be an integer division, right?

Reply:
~/ is integer division in dart. / produces double by default:

https://dart.dev/guides/language/language-tour#arithmetic-operators